nemo/collections/nlp/models/machine_translation/mt_enc_dec_model.py

```diff
@@ -0,0 +1,317 @@
+from sacrebleu import corpus_bleu
+
+from nemo.collections.common.losses import SmoothedCrossEntropyLoss
+from nemo.collections.common.metrics import Perplexity
+from nemo.collections.common.parts import transformer_weights_init
+from nemo.collections.nlp.data import TranslationDataset
+from nemo.collections.nlp.data.machine_translation.machine_translation_dataset import TranslationDataConfig
+from nemo.collections.nlp.models.enc_dec_nlp_model import EncDecNLPModel, EncDecNLPModelConfig
+from nemo.collections.nlp.modules.common.transformer import BeamSearchSequenceGenerator
+from nemo.core.classes.common import typecheck
+from nemo.utils import logging
+
+__all__ = ['MTEncDecModel']
+
+
+@dataclass
+class MTEncDecModelConfig(EncDecNLPModelConfig):
+    train_ds: Optional[TranslationDataConfig] = None
+    validation_ds: Optional[TranslationDataConfig] = None
+    test_ds: Optional[TranslationDataConfig] = None
+    num_val_examples: int = 3
+    num_test_examples: int = 3
+    beam_size: int = 1
+    len_pen: float = 0.0
+    max_generation_delta: int = 50
+    label_smoothing: Optional[float] = 0.0
+    vocab_divisible_by_eight: bool = True
+
+
+class MTEncDecModel(EncDecNLPModel):
+    """
+    Left-to-right Transformer language model.
+    """
+
+    def __init__(self, cfg: MTEncDecModelConfig, trainer: Trainer = None):
+
+        self.setup_enc_dec_tokenizers(cfg)
+
+        super().__init__(cfg=cfg, trainer=trainer)
+
+        # make vocabulary size divisible by 8 for fast fp16 training
+        if cfg.vocab_divisible_by_eight:
+            self.encoder_vocab_size = 8 * math.ceil(self.encoder_tokenizer.vocab_size / 8)
+            self.decoder_vocab_size = 8 * math.ceil(self.decoder_tokenizer.vocab_size / 8)
+            cfg.encoder_embedding.vocab_size = self.encoder_vocab_size
+            cfg.decoder_embedding.vocab_size = self.decoder_vocab_size
+
+        self.encoder_embedding = instantiate(cfg.encoder_embedding)
+        self.decoder_embedding = instantiate(cfg.decoder_embedding)
+        self.encoder = instantiate(cfg.encoder)
+        self.decoder = instantiate(cfg.decoder)
+
+        # TODO: Optionally tie weights
+
+        self.log_softmax = instantiate(cfg.head)
+
+        self.beam_search = BeamSearchSequenceGenerator(
+            embedding=self.decoder_embedding,
+            decoder=self.decoder,
+            log_softmax=self.log_softmax,
+            max_sequence_length=cfg.decoder_embedding.max_sequence_length,
+            beam_size=cfg.beam_size,
+            bos=self.decoder_tokenizer.bos_id,
+            pad=self.decoder_tokenizer.pad_id,
+            eos=self.decoder_tokenizer.eos_id,
+            len_pen=cfg.len_pen,
+            max_delta_length=cfg.max_generation_delta,
+        )
+
+        # TODO: encoder and decoder with different hidden size?
+        std_init_range = 1 / cfg.encoder.hidden_size ** 0.5
+        self.apply(lambda module: transformer_weights_init(module, std_init_range))
```

Comment:
this line should be last line of the constructor

Reply:
Done.